core/src/main/java/com/linecorp/armeria/client/DnsResolverGroupBuilder.java

```diff
@@ -0,0 +1,344 @@
+import io.netty.resolver.AddressResolver;
+import io.netty.resolver.AddressResolverGroup;
+import io.netty.resolver.HostsFileEntriesResolver;
+import io.netty.resolver.ResolvedAddressTypes;
+import io.netty.resolver.dns.DefaultDnsCnameCache;
+import io.netty.resolver.dns.DnsCnameCache;
+import io.netty.resolver.dns.DnsNameResolver;
+import io.netty.resolver.dns.DnsNameResolverBuilder;
+import io.netty.resolver.dns.DnsQueryLifecycleObserverFactory;
+import io.netty.resolver.dns.DnsServerAddressStreamProvider;
+import io.netty.resolver.dns.NoopAuthoritativeDnsServerCache;
+import io.netty.resolver.dns.NoopDnsCache;
+
+/**
+ * Builds an {@link AddressResolverGroup} which builds {@link AddressResolver}s that update DNS caches
+ * automatically. A usual DNS resolver such as {@link DnsNameResolver}, a cache is expired after TTL.
+ * On the other hand, the refreshing {@link AddressResolver} updates the DNS cache spontaneously
+ * even after TTL has elapsed. If refreshing fails, the {@link AddressResolver} will retry with
+ * {@link #refreshBackoff(Backoff)} until {@link #refreshTimeoutMillis(long)} has elapsed.
+ *
+ * <p>{@link AddressResolver} keeps cache hits so that it does not refresh the caches for
+ * the DNS addresses used only once.
+ */
+public final class DnsResolverGroupBuilder {
+
+    private Backoff refreshBackoff = Backoff.ofDefault();
+
+    private long refreshTimeoutMillis = TimeUnit.MINUTES.toMillis(10);
+
+    private int minTtl = 1;
+    private int maxTtl = Integer.MAX_VALUE;
+
+    // DnsNameResolverBuilder properties:
+
+    private boolean traceEnabled = true;
+    @Nullable
+    private DnsCnameCache cnameCache;
+    @Nullable
+    private Long queryTimeoutMillis;
+    @Nullable
+    private ResolvedAddressTypes resolvedAddressTypes;
+    @Nullable
+    private Boolean recursionDesired;
+    @Nullable
+    private Integer maxQueriesPerResolve;
+    @Nullable
+    private Integer maxPayloadSize;
+    @Nullable
+    private Boolean optResourceEnabled;
+    @Nullable
+    private HostsFileEntriesResolver hostsFileEntriesResolver;
+    @Nullable
+    private DnsServerAddressStreamProvider dnsServerAddressStreamProvider;
+    @Nullable
+    private DnsQueryLifecycleObserverFactory dnsQueryLifecycleObserverFactory;
+    @Nullable
+    private List<String> searchDomains;
+    @Nullable
+    private Integer ndots;
+    @Nullable
+    private Boolean decodeIdn;
+
+    DnsResolverGroupBuilder() {}
+
+    /**
+     * Sets {@link Backoff} which is used when the {@link DnsNameResolver} fails to update the cache.
+     */
+    public DnsResolverGroupBuilder refreshBackoff(Backoff refreshBackoff) {
+        this.refreshBackoff = requireNonNull(refreshBackoff, "refreshBackoff");
+        return this;
+    }
+
+    /**
+     * Sets the timeout of refreshing DNS caches. After this timeout, the previously resolved
+     * DNS cache is invalidated.
+     */
+    public DnsResolverGroupBuilder refreshTimeout(Duration refreshTimeout) {
```

Comment:
Is this needed on top of the backoff's max attempts?

Reply:
I think so. Because there's jitter which makes the user hard to calculate the whole duration, I thought having `refreshTimeout` is better so the cache stays until TTL + `refreshTimeout`. Do you think we need to replace it or add  `max attempts`?

Reply:
`refreshTimeout` seems to be relatively unscientific anyways (default is a large 10 min) so not sure if it matters so much whether the duration is precise with respect to jitter. On the flip side, the interaction between max attempts and refreshTimeout is confusing since two different variables affect the same behavior.

I feel that if we do want to add a hard max deadline on retries, it's better if we add it as another parameter `Backoff` than here, so at least it's kept together with max attempts. I think the similar concern can apply to other uses of `Backoff`.

Reply:
Ah, I forgot about that we have `Backoff.withMaxAttempts()` so we might not need to add it explicitly. 🤣 So we can just remove the `refreshTimeout` here if we don't need it.